refresh the access token in the background

BacLuc opened the pull request:
Use gesdinet/jwt-refresh-token-bundle to achieve this. The access token lifetime was reduced to 30 mins
according to https://cloud.google.com/apigee/docs/api-platform/antipatterns/oauth-long-expiration as a starting point.
Do not log out if the refresh in the background fails, the user might want to backup the content he is currently editing.

One proposed pattern is to intercept 401 responses and refresh the token when the requests fail. Because we often have multiple requests running, this did not work well. Now we have a timeout in the background which refreshes the access token periodically.

closes #4898

BacLuc committed:
refresh the access token in the background

Use gesdinet/jwt-refresh-token-bundle to achieve this.
The access token lifetime was reduced to 30 mins
according to https://cloud.google.com/apigee/docs/api-platform/antipatterns/oauth-long-expiration
as a starting point.
Do not log out if the refresh in the background fails, the user might want to backup
the content he is currently editing.

One proposed pattern is to intercept 401 responses and refresh the token when the requests fail.
Because we often have multiple requests running, this did not work well.
Now we have a timeout in the background which refreshes the access token periodically.

closes #4898

carlobeltrame reviewed (changes requested):
This did not work in my tests yet.
- I logged in with username and password, and got 3 cookies
- I waited for more than 30 minutes
- I refreshed the page
- I got to the login page and suddenly had 4 cookies. The fourth one called ph_phc_Tbfg4EiRsr5iefEoth2Y1Hi3sttTeLQ5RV5TLg4hL1W_posthog. What is this new cookie for?
- The terms of service on the landing page also need to be adjusted, right now they state that all cookies are deleted after 12 hours. The refresh token I got was valid for 1 month.

BacLuc commented:
@codeRabbit review this PR

BacLuc commented on api/composer.json:
```diff
@@ -15,6 +15,7 @@
         "exercise/htmlpurifier-bundle": "5.1",
         "friendsofsymfony/http-cache": "3.1.1",
         "friendsofsymfony/http-cache-bundle": "3.2.0",
+        "gesdinet/jwt-refresh-token-bundle": "1.5.0",
```
There are no published 2.0 releases.

BacLuc commented on api/config/packages/gesdinet_jwt_refresh_token.yaml:
```diff
@@ -0,0 +1,11 @@
+gesdinet_jwt_refresh_token:
```
The default of this package for the refresh token is 1 month.
The article of google said 24h which seemed a little strict for an application you don't use every day. https://cloud.google.com/apigee/docs/api-platform/antipatterns/oauth-long-expiration
I did not find other recommendations.
(But i read somewhere that there are applications which even use 90 days)

What do you say @carlobeltrame?

BacLuc replied:
We start by not reducing it and making it configurable.

BacLuc commented on frontend/src/plugins/auth.js:
```diff
@@ -11,6 +11,51 @@
+  )
+}
+
+async function refreshAndSchedule() {
+  await refresh()
+  rescheduleRefresh()
+}
```
If refresh fails, rescheduleRefresh is not called -> no refresh anymore.

manuelmeister commented:
## Core Meeting Decision

Nice work!
Please leave the lifetime as is (for the moment, until we have a smooth refresh for all browsers)
Make TTL configurable

simfeld commented:
Minor: when you sign out, delete local storage and then refresh the login page you get this error

<img width="774" height="302" alt="image" src="https://github.com/user-attachments/assets/c127a10d-6812-42fa-988d-8ce72e175767" />

BacLuc commented on api/config/routes.yaml:
```diff
@@ -4,3 +4,5 @@
 authentication_token:
   path: /authentication_token
   methods: ['POST']
+api_refresh_token:
+  path: /token/refresh
```
done

BacLuc commented on api/config/services.yaml:
```diff
@@ -103,6 +103,12 @@ services:
+      arguments:
+        - '@.inner'
+        - '%env(COOKIE_PREFIX)%'
+
```
done

BacLuc commented on api/src/OpenApi/RefreshTokenDecorator.php:
```diff
@@ -0,0 +1,44 @@
+                        'description' => "Get a refreshed JWT token split across the two cookies {$cookiePrefix}jwt_hp and {$cookiePrefix}jwt_s. Also returns a new refresh token {$cookiePrefix}refresh_token.",
+                    ],
+                ],
+                summary: 'Refresh token.',
+                requestBody: new Model\RequestBody(
+                    description: 'Body',
+                    content: new \ArrayObject([
+                        'application/ld+json' => [
+                            'schema' => [],
+                        ],
+                        'application/json' => [
+                            'schema' => [],
+                        ],
+                    ]),
+                ),
```
done

BacLuc commented on frontend/src/plugins/__tests__/auth.spec.js:
```diff
@@ -340,6 +347,176 @@
+    vi.spyOn(apiStore, 'href').mockImplementation(async (input) => {
+      return input
+    })
+  })
```
I will add the tests later.

BacLuc commented on frontend/src/plugins/auth.js:
```diff
@@ -11,6 +11,51 @@ axios.interceptors.response.use(null, (error) => {
+    refreshAndSchedule,
+    (getJWTExpirationTimestamp() - Date.now()) / 2
+  )
+}
```
I will add the tests later.

BacLuc commented on frontend/src/plugins/auth.js:
```diff
@@ -11,6 +11,51 @@
+async function refresh() {
+  const url = apiStore.href(apiStore.get(), 'refreshToken')
+  return apiStore.post(url)
+}
```
I will add the tests later.

BacLuc commented on frontend/src/plugins/__tests__/auth.spec.js:
```diff
@@ -340,6 +347,176 @@
+        'localhost_jwt_hp',
+        createJWT({ exp: new Date().getTime() + expiresInMs })
+      )
+    })
```
I will add the tests later.

BacLuc commented on api/tests/Api/SnapshotTests/__snapshots__/ResponseSnapshotTest__testOpenApiSpecMatchesSnapshot__1.yml:
```diff
@@ -34692,6 +34692,24 @@ paths:
+          application/ld+json:
+            schema: []
+        description: Body
+        required: false
```
done

BacLuc commented:
Github somehow cannot show the checks that are running.
For commit https://github.com/ecamp/ecamp3/pull/7942/commits/6e1296eec705b95644c6637818a68ed41ada8ed3
you find the checks here: https://github.com/BacLuc/ecamp3/actions/runs/17194309791

BacLuc committed:
api: make AUTHENTICATION_TOKEN_TTL configurable

Don't add it to the workflows yet because https://github.com/ecamp/ecamp3/pull/8025 will make this a lot easier.
Has merge conflict with https://github.com/ecamp/ecamp3/pull/8025

The ttl will be made shorter ones all users have refresh tokens.

BacLuc committed:
frontend: perform token refresh on visibilityChange to visible

That you also refresh the token when you reactivate the tab.
Using the vuex store to get this value was not reliable
when the tab was reactivated.
-> get it directly from localStorage

Issue: #4898

BacLuc committed:
frontend: perform token refresh when you have connectivity again

Issue: #4898

BacLuc commented:
> ## Core Meeting Decision
> 
> Nice work! Please leave the lifetime as is (for the moment, until we have a smooth refresh for all browsers) Make TTL configurable

Made the TTL configurable up to the helm chart.
Waiting for #8025, this makes the last step way easier.

carlobeltrame reviewed (changes requested):
It works a little better now, but I still have some issues:

- I logged in with username and password, and got 3 cookies
- I noticed that the session cookies had an expiry time for 12 hours. I didn't want to wait that long, so I deleted the jwt_hp and jwt_s cookies manually (to simulate them expiring)
- When refreshing the page in this state, I get to the login page for a split second and then back to where I was before, with 3 cookies. This works correctly :+1: 
- When doing any interaction other than refreshing the page instead (in this state with expired session cookies), I just get sent to the login page. This includes navigating any link in the app, or trying to change some input field.

Since the normal interactions with eCamp v3 are not to just refresh the page, I think this is not really helping yet.

Also, we still need an adjustment PR to the terms of service, as mentioned in my previous comment.

BacLuc commented:
> It works a little better now, but I still have some issues:
> 
>     * I logged in with username and password, and got 3 cookies
> 
>     * I noticed that the session cookies had an expiry time for 12 hours. I didn't want to wait that long, so I deleted the jwt_hp and jwt_s cookies manually (to simulate them expiring)
> 
>     * When refreshing the page in this state, I get to the login page for a split second and then back to where I was before, with 3 cookies. This works correctly 👍
> 
>     * When doing any interaction other than refreshing the page instead (in this state with expired session cookies), I just get sent to the login page. This includes navigating any link in the app, or trying to change some input field.
> 
> 
> Since the normal interactions with eCamp v3 are not to just refresh the page, I think this is not really helping yet.
> 
> Also, we still need an adjustment PR to the terms of service, as mentioned in my previous comment.

Could you please test the same i did?
The description does not help me reproduce your issues. (when do you delete the cookie, how do you delete the cookie)
I reduced the ttl again, maybe this helps you testing it on the pr deployment.

BacLuc commented:
Sorry, the update did not inject the correct lifetime of the cookie. I will fix that.

BacLuc commented:
> Sorry, the update did not inject the correct lifetime of the cookie. I will fix that.

Now the short ttl is active, fixed it

carlobeltrame commented on frontend/src/plugins/auth.js:
```diff
@@ -11,6 +12,55 @@ axios.interceptors.response.use(null, (error) => {
+}
+
+async function refreshAndSchedule() {
+  console.log('executing refresh')
```
Remove? Or do you want to keep it for now and remove once we know it works?

BacLuc replied:
done

carlobeltrame reviewed (approved):
Looks great, thanks!
